### PR TITLE
settings: assert that SystemOnly settings are not accessed in virtual clusters

### LINKED ISSUE #91825: *: reads of SystemOnly settings from tenant process
**Describe the problem**

SystemOnly settings are not intended to be read from a tenant process.  However, nothing actually disallows a tenant process from reading the in-memory default of a SystemOnly setting and taking actions on it.

The settings implementation includes a safety check for this was intended to be enabled in test builds, but that safety check requires `(*Values).SetNonSystemTenant` is called at some point during tenant startup.  Currently we do not call that method so no accesses to SystemOnly settings from tenants are disallowed.

This recently tripped us up here https://github.com/cockroachdb/cockroach/pull/91824 and a quick test run with SetNonSystemTenant immediately hit our safety check on `scheduler_latency.sample_period`.  It is likely that there are a number of other SystemOnly settings being accessed from tenant processes.

We should enable SetNonSystemTenant in our tests and work through the problems.

Jira issue: CRDB-21443
Epic: CRDB-6671

### PR DESCRIPTION
**Pull request by knz**
Fixes #91825.
Epic: CRDB-6671
First commit from #110947

TLDR: tests will now fail (with a panic) if code running in virtual clusters mistakenly attempts to access a SystemOnly cluster setting.

----

Prior to this patch, it was possible for code (e.g. in SQL) running inside a virtual cluster to access a `SystemOnly` setting. This resulted in silently incorrect behavior: the default value of the setting would be observed always, without any relationship to values explicitly set in the storage layer.

When the setting mechanisms for virtual clusters had been implemented orignally, this concern was known. We even had implemented a mechanism on `settings.Values` (`SetNonSystemTenant()`) by which we could cause further `.Get()` calls to fail with an error in tests if accessing a `SystemOnly` setting from a virtual cluster.

Sadly, this mechanism was never hooked up anywhere.

This patch closes that loop by annotating the `settings.Values` properly during server initialization.

Since the property is stored inside the `settings.Values`, this means it is not any more possible to share a single `settings.Values` instance (nor `cluster.Settings`) between the system tenant and a secondary tenant.

----

The following bugs were fixed as a result of this change:

- certain settings didn't have the right class, including:

  - kv.bulk_io_write.concurrent_export_requests
  - kv.closed_timestamp.follower_reads.enabled
  - kv.range_split.by_load_merge_delay
  - kv.rangefeed.enabled
  - sql.hash_sharded_range_pre_split.max
  - storage.max_sync_duration.fatal.enabled
  - storage.value_blocks.enabled
  - ui.display_timezone
  - kv.bulk_sst.target_size

- the API endpoint that lists cluster settings was incorrectly
  trying to retrieve SystemOnly settings when accessed
  from a virtual cluster.

### CODE REVIEW

**Comment by knz** [comment]
Quite some damage from this assertion... 
image

**Comment by RaduBerinde** [comment]
Hm, I thought we did this early on, we were panicking in test builds (https://github.com/cockroachdb/cockroach/pull/74253). I wonder what changed.

**Comment by knz** [comment]
> Hm, I thought we did this early on, we were panicking in test builds (https://github.com/cockroachdb/cockroach/pull/74253). I wonder what changed.

Unless I'm mistaken, this PR wasn't complete. The `SetNonSystemTenant()` method wasn't called anywhere.

**Comment by RaduBerinde** [comment]
> Unless I'm mistaken, this PR wasn't complete. The `SetNonSystemTenant()` method wasn't called anywhere.

Wow, you're right, I'm such a hack haha

**Review by yuzefovich — COMMENTED** [review]
Only looked at the last commit, looks good to me.

> Reviewed 55 of 55 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-:-NedjBS_1kSu5Wvjtolv:b-jig3qc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @stevendanna)

___
*[`pkg/server/testserver.go` line 600 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedfbbp3B7Da3zUCAFd:-Nedfbbp3B7Da3zUCAFe:b-2hhm8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/server/testserver.go#L600)):*
> ```Go
> 		// - call serverutils.StartServer
> 		// - expect the overrides to propagate to the application layer.
> 		tenantSettings.SV.TestingCopyForVirtualCluster(&st.SV)
> ```

This is nice, I like it a lot. However, are there cases where this behavior is undesirable? Should we introduce a knob that disables this? Or perhaps do that only when / if such need arises? (I'm leaning towards the latter.)
___
*[`pkg/settings/values.go` line 251 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedg_UY-ueGDvz0VWm-:-Nedg_UY-ueGDvz0VWm0:bmbhcix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/settings/values.go#L251)):*
> ```Go
> // TestingCopyForVirtualCluster makes a copy of the input Values in
> // the target Values for use when initializing a server for a virtual
> // cluster in tests. This is meant to propagate overrides
> ```

I'm confused a bit here - why are we not copying TenantReadOnly? Is it because the underlying container for each RO setting is directly shared between different Values?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110676)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @stevendanna, and @yuzefovich)

___
*[`pkg/settings/values.go` line 251 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedg_UY-ueGDvz0VWm-:-NejEScA1pL9d8P64Z31:b-5r0kej) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/settings/values.go#L251)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I'm confused a bit here - why are we not copying TenantReadOnly? Is it because the underlying container for each RO setting is directly shared between different Values?
</blockquote></details>

We just discussed this - we need this to propagate.
Alas the other work in #110758 is only going to work with persisted changes (they will be picked up by rangefed). So we need a propagate here for the .Override changes.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110676)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @stevendanna, and @yuzefovich)

___
*[`pkg/server/testserver.go` line 600 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedfbbp3B7Da3zUCAFd:-Nekkfuc6hp3nd-mlE_w:b-6oe1z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/server/testserver.go#L600)):*
> Or perhaps do that only when / if such need arises?

I like that better too.
___
*[`pkg/settings/values.go` line 251 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedg_UY-ueGDvz0VWm-:-NekkjouA7fWsdVaugs_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/settings/values.go#L251)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

We just discussed this - we need this to propagate.
Alas the other work in #110758 is only going to work with persisted changes (they will be picked up by rangefed). So we need a propagate here for the .Override changes.
</blockquote></details>

Done.

**Review by yuzefovich — APPROVED** [review]
Last commit :lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-:-Nemz2Rp7zvhVmQD2cNv:b5tyyy8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @knz, and @stevendanna)

___
*[`pkg/settings/values.go` line 251 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedg_UY-ueGDvz0VWm-:-NemiokX93b8PCPirgsb:b-tjx9kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/settings/values.go#L251)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: update the comment too to mention RO.
___
*[`pkg/storage/pebble.go` line 83 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-NemkAyNBy2w1Ll7f8BA:-NemkAyNBy2w1Ll7f8BB:b-5aob0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6733ad7d7902c9580634260a564fb7a01729749f/pkg/storage/pebble.go#L83)):*
> ```Go
> // MaxSyncDuration fatal the Cockroach process. Defaults to true.
> var MaxSyncDurationFatalOnExceeded = settings.RegisterBoolSetting(
> 	settings.TenantWritable, // used for temp storage in virtual cluster servers.
> ```

nit: usually for inline comments like this we start without a capital and do not have a period in the end.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110676)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @stevendanna, and @yuzefovich)

___
*[`pkg/settings/values.go` line 251 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-Nedg_UY-ueGDvz0VWm-:-Nen-oOP9BoDSVXJToGM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a1e96020bebbe70796ba2126fe36345d81392b8/pkg/settings/values.go#L251)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: update the comment too to mention RO.
</blockquote></details>

Done.
___
*[`pkg/storage/pebble.go` line 83 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/110676#-NemkAyNBy2w1Ll7f8BA:-Nen-wT3DdMrMAeVyY__:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6733ad7d7902c9580634260a564fb7a01729749f/pkg/storage/pebble.go#L83)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: usually for inline comments like this we start without a capital and do not have a period in the end.
</blockquote></details>

Fixed.

**Comment by knz** [comment]
bors r=yuzefovich

**Comment by yuzefovich** [comment]
Second commit should be removed, right?

bors r-

**Comment by knz** [comment]
> Second commit should be removed, right?

:facepalm: 

Done

bors r=yuzefovich
